src/realm.h

```diff
@@ -1559,37 +1559,271 @@ RLM_API void realm_collection_changes_get_ranges(
+ * not report an error).
+ *
+ * @param value The value to insert.
+ * @param out_index If non-null, will be set to the index of the inserted
+ *                  element, or the index of the existing element.
+ * @param out_inserted If non-null, will be set to true if the element did not
+ *                     already exist in the set. Otherwise set to false.
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_set_insert(realm_set_t*, realm_value_t value, size_t* out_index, bool* out_inserted);
+
+/**
+ * Erase an element from a set.
+ *
+ * If the element does not exist in the set, this function does nothing (and
+ * does not report an error).
+ *
+ * @param value The value to erase.
+ * @param out_erased If non-null, will be set to true if the element was found
+ *                   and erased, and otherwise set to false.
+ * @return True if no exception occurred.
+ */
 RLM_API bool realm_set_erase(realm_set_t*, realm_value_t value, bool* out_erased);
+
+/**
+ * Clear a set of values.
+ *
+ * @return True if no exception occurred.
+ */
 RLM_API bool realm_set_clear(realm_set_t*);
-RLM_API bool realm_set_assign(realm_set_t*, realm_value_t values, size_t num_values);
-RLM_API realm_notification_token_t* realm_set_add_notification_callback(realm_object_t*, void* userdata,
+
+/**
+ * In a set of objects, delete all objects in the set and clear the set. In a
+ * set of values, clear the set.
+ *
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_set_remove_all(realm_set_t*);
+
+/**
+ * Replace the contents of a set with values.
+ *
+ * The provided values may contain duplicates, in which case the size of the set
+ * after calling this function will be less than @a num_values.
+ *
+ * @param values The list of values to insert.
+ * @param num_values The number of elements.
+ * @return True if no exception occurred.
+ */
+RLM_API bool realm_set_assign(realm_set_t*, const realm_value_t* values, size_t num_values);
+
+/**
+ * Subscribe to notifications for this object.
+ *
+ * @return A non-null pointer if no exception occurred.
+ */
+RLM_API realm_notification_token_t* realm_set_add_notification_callback(realm_set_t*, void* userdata,
                                                                         realm_free_userdata_func_t free,
                                                                         realm_on_collection_change_func_t on_change,
                                                                         realm_callback_error_func_t on_error,
                                                                         realm_scheduler_t*);
 
+/**
+ * Get a dictionary instance for the property of an object.
+ *
+ * Note: It is up to the caller to call `realm_release()` on the returned dictionary.
+ *
+ * @return A non-null pointer if no exception occurred.
+ */
+RLM_API realm_dictionary_t* realm_get_dictionary(realm_object_t*, realm_property_key_t);
```

Comment:
Are we only targeting support for string based keys for now? At least I don't see options for specifying the key type of dicts.